Fix flashlight not always matching gameplay scaling

frenzibyte opened the pull request:
- Closes #34703 

This fix essentially expands on https://github.com/ppy/osu/pull/33910 by extracting scale directly from the playfield rather than just reading `PlayfieldAdjustmentContainer.Scale` (which does not cover any scaling factor applied in nested containers between the playfield adjustment container and the playfield itself).

This also required size adjustments in osu! and catch, as well as a minor change in taiko flashlight:
 - In osu!, because I'm now scaling the flashlight based on the playfield, there's a constant 1.6x factor applied directly by `OsuPlayfieldAdjustmentContainer.ScalingContainer`. To keep the flashlight size from visually getting 1.6x larger, I've adjusted the default flashlight size to counter that. Applied in https://github.com/ppy/osu/commit/29b2a7167ecbd1db4b97e2510006b47180a582f4.
 - In osu!catch, same as osu!, there's a 1.6x factor applied by `CatchPlayfieldAdjustmentContainer.ScalingContainer`. Also applied in https://github.com/ppy/osu/commit/29b2a7167ecbd1db4b97e2510006b47180a582f4.
 - In osu!taiko, since https://github.com/ppy/osu/pull/33910 was merged, the flashlight size became scaled twice by the playfield, once in:

   https://github.com/ppy/osu/blob/30f7da8f716a48f0c3ebbea3b1f95a50405cc75a/osu.Game.Rulesets.Taiko/Mods/TaikoModFlashlight.cs#L66 and again in:
  https://github.com/ppy/osu/blob/30f7da8f716a48f0c3ebbea3b1f95a50405cc75a/osu.Game/Rulesets/Mods/ModFlashlight.cs#L159-L166

   This made the flashlight grow its view as the playfield widens:

   https://github.com/user-attachments/assets/e4da4e24-f656-4e13-af23-19e26cc35eea

   This is fixed in https://github.com/ppy/osu/commit/5a84c076fcc271c0c6f401435249498175638fd6 by removing the local logic in favour of the main scaling logic added by this PR in the flashlight draw node. PR:

   

   https://github.com/user-attachments/assets/6a5f47f4-5c1a-4b3e-82db-2241359c8420

   For comparison, this is how the flashlight size is with just the changes in https://github.com/ppy/osu/pull/33910 reverted (point is to show that it matches what I have in this PR, i.e. the correct behaviour):

   https://github.com/user-attachments/assets/17d0877a-3132-4f56-9606-b1c9ab0889ee

 - In osu!mania, no direct changes were made, but it seems this PR fixes an issue with flashlight that has existed since mobile mania was improved. Before:


   https://github.com/user-attachments/assets/970ffbc8-034c-4800-8039-bd77152d8561

   After:

   https://github.com/user-attachments/assets/82a54c44-e301-46e8-85ad-4ce02b46c656

frenzibyte committed:
Fix flashlight not correctly scaled to match playfield

frenzibyte committed:
Adjust default flashlight size on osu! & osu!catch

Because the flashlight is made to be scaled by playfield, there are
constant scale factors applied somewhere in the
`PlayfieldAdjustmentContainer` which needs to be reflected in the
flashlight size to keep the size the same.

The factor is specifically 1.6x, computed in {Osu,Catch}PlayfieldAdjustmentContainer.ScalingContainer`.

More generally, I've deduced these factors by logging the difference
between the `flashlightSize` before and after b78abe2f.

frenzibyte committed:
Fix taiko flashlight applying aspect ratio twice

frenzibyte committed:
Add visual test setup for taiko flashlight

peppy commented:
@frenzibyte test failures look relevant

bdach reviewed (commented):
I would expect screenshots with cross-comparison with stable (and lazer master) in this PR if the flashlight constants are being touched, please provide them

peppy commented on osu.Game.Rulesets.Catch/Mods/CatchModFlashlight.cs:
```diff
@@ -24,7 +24,7 @@ public partial class CatchModFlashlight : ModFlashlight<CatchHitObject>
         public override BindableBool ComboBasedSize { get; } = new BindableBool(true);
 
-        public override float DefaultFlashlightSize => 325;
+        public override float DefaultFlashlightSize => 203.125f;
```
100% has to be a constant, hopefully in the place its primarily used.

frenzibyte replied:
I meant to ask about the current numbers of the `DefaultFlashlightSize` (200 in osu! and 325 in catch before this PR), if they're not referenced anywhere else and they only represent themselves, I'm not sure it's necessary to split the constant away from them rather than just inline the computation?

frenzibyte committed:
Remove no longer valid test

frenzibyte committed:
Fix flashlight not handling internal playfield sizing changes

Note that this does not handle sizing/scaling changes applied directly
to `Playfield`, but it handles any changes within the layers inside
`PlayfieldAdjustmentContainer`.

frenzibyte committed:
Fix xmldoc error and reword

bdach reviewed (approved):
probably fine
